File: arrays/two-number-sum.py
# time: O(n^2)
# space: O(1)
# where n is the length of the input array
def twoNumberSum1(array, targetSum):
    for i in range(len(array) - 1):
		firstNum = array[i]
		for j in range(i + 1, len(array)):
			  secondNum = array[j]
			  if firstNum + secondNum == targetSum:
				    return sorted([firstNum, secondNum])
	  return []


# time: O(n)
# space: O(n)
# where n is the length of the input array
def twoNumberSum2(array, targertSum):
    nums = {}
    for num in array:
        potentialMatch = targertSum - num
        if potentialMatch in nums:
            return sorted([potentialMatch, num])
        else:
            nums[num] = True
    return []


# time: O(n(log(n)))
# space: O(1)
# where n is the length of the input array
def twoNumberSum3(array, targetSum):
    array.sort()
    left = 0
    right = len(array) - 1
    while left < right:
        currentSum = array[left] array[right]
        if currentSum == targetSum:
            return [array[left], array[right]]
        elif currentSum < targetSum:
            left += 1
        elif currentSum > targetSum:
            right -= 1
    return []


twoNumberSum1([3, 5, -4, 8, 11, 1, -1, 6], 10) #[-1, 11]

File: arrays/two-number-sum.js
// time: O(n)
// space: O(n)
const twoNumberSum1 = (array, targetSum) => {
  const cache = {}

  for (let i = 0; i < array.length; i++) {
    const val = targetSum - array[i]

    if (array[i] in cache) {
      return [Math.min(array[i], cache[array[i]]), Math.max(array[i], cache[array[i]])]
    } else {
      cache[val] = array[i]
    }
  }

  return []
}

// time: O(nlog(n))
// space: O(1)
const twoNumberSum2 = (array, targetSum) => {
  array.sort((a, b) => a - b)
  let left = 0
  let right = array.length - 1

  while (left < right) {
    if (array[left] + array[right] > targetSum) {
      right--
    } else if (array[left] + array[right] < targetSum) {
      left++
    } else {
      return [array[left], array[right]]
    }
  }

  return []
}

console.log(twoNumberSum1([-21, 301, 12, 4, 65, 56, 210, 356, 9, -47], 163)) // [-47, 210]
